doc/opdatasigverify.md

```diff
@@ -0,0 +1,118 @@
+
+OP_DATASIGVERIFY looks at this signature type byte to determine the signature validation algorithm and then executes that algorithm.  There is currently one defined algorithm:
+
+#### value 1: DATASIG_COMPACT_ECDSA
+
+This algorithm is the same operation as the Bitcoin Cash message verification RPC call (e.g signmessage/verifymessage).  The signature must therefore be a pubkey-recoverable signature encoded in bitcoin's "compact" format.  The authoritative specification of this signature is beyond the scope of this document, but in essence it is comprised of a 1 byte recovery ID, 32 bytes "r", 32 bytes "s".
+
+The algorithm first computes the double-sha256 hash of the byte string "Bitcoin Signed Message:\n" prepended to the supplied *data* (stack top - 2).  It then computes a pubkey from this hash and the provided signature (stack top - 1, without the first byte).  If the pubkey cannot be recovered, the script fails.  It then compares the hash160 (same as OP_HASH160) of the recovered pubkey to the provided *pubkeyhash* (stack top).  If the comparison fails, the script fails.
+
+Otherwise, the top 2 items are popped off the stack (leaving *data* on the top of the stack), and the opcode succeeds.
+
+## Examples
+
+* Simplest DATASIGVERIFY:
+  * output script:  `pubkeyhash OP_DATASIGVERIFY`
+  * spend script: `data sig`
+
+* Realistic example: Verify that `data` is signed by `dpubkeyhash` (signature `daddrsig`) and equals `x`.  Also do normal p2pkh transaction verification on `pubkeyhash` (with signature `txSig` and public key `pubkey`):
+  * output script: `dpubkeyhash OP_DATASIGVERIFY x OP_EQUALVERIFY OP_DUP OP_HASH160 pubkeyhash OP_EQUALVERIFY OP_CHECKSIG
+  * spend script: `txSig pubkey data daddrsig`
+
+## Discussion
+
+### Why is the non-verify version (e.g. OP_CHECKDATASIG) unnecessary?
+
+An instruction like OP_CHECKDATASIG would validate the signature on data and push a true/false result on the stack, analguous to OP_CHECKSIG vs OP_CHECKSIGVERIFY.
+
+A Bitcoin Cash script is not exactly like a normal program.  Its fundamental and only purpose is to encumber spendability -- your input script must meet the output script's requirements in order to spend the coins.  Therefore, any program sequence that does not narrow the group of possible spenders is equivalent to a simple choice.  In the OP_CHECKDATASIG case, any potential spender can create a spend script that fails OP_CHECKDATASIG by pushing random bytes rather than a valid signature, so the "false" result is equivalent to (but less efficient than) an if statement on a pushed constant.
+
+Formally, any script with encumberances x, y, z and solutions x', y', and z', of the form:
+
+`pubkeyhash OP_CHECKDATASIG OP_IF x OP_ELSE y OP_ENDIF z`
```

Comment:
Is `OP_CHECKDATASIG` the same as `OP_DATASIGVERIFY` ?

Reply:
no, OP_CHECKDATASIG is a hypothetical additional instruction that doesn't do the VERIFY part.  Just like there's an OP_CHECKSIG and OP_CHECKSIGVERIFY.  I am explaining why it is not necessary.